src/java.base/share/classes/java/util/stream/GathererOp.java

```diff
@@ -0,0 +1,745 @@
+            if (isEmpty())
+                return that;
+
+            if (!that.isEmpty()) {
+                final var tb = that.build();
+                if (rightMost != null && tb instanceof NodeBuilder.Builder<X>
+                && tb.count() < 8) // TODO revisit cutoff point
+                    tb.forEach(this); // Avoid conc for small nodes
+                else
+                    leftMost = Nodes.conc(StreamShape.REFERENCE, this.build(), tb);
+            }
+
+            return this;
+        }
+
+        public Node<X> build() {
+            if (isEmpty())
+                return Nodes.emptyNode(StreamShape.REFERENCE);
+
+            final var rm = rightMost;
+
+            if (rm != null) {
+                rightMost = null; // Make sure builder isn't reused
+                final var lm = leftMost;
+                leftMost = (lm == null) ? rm : Nodes.conc(StreamShape.REFERENCE, lm, rm);
+            }
+
+            return leftMost;
+        }
+    }
+
+    final static class GatherSink<T, A, R> implements Sink<T>, Gatherer.Downstream<R> {
+        private final Sink<R> sink;
+        private final Gatherer<T, A, R> gatherer;
+        private final Integrator<A, T, R> integrator; // Optimization: reuse
+        private A state;
+        private boolean proceed = true;
+
+        GatherSink(Gatherer<T, A, R> gatherer, Sink<R> sink) {
+            this.gatherer = gatherer;
+            this.sink = sink;
+            this.integrator = gatherer.integrator();
+        }
+
+        // java.util.stream.Sink contract below:
+
+        @Override
+        public void begin(long size) {
+            final var initializer = gatherer.initializer();
+            if (initializer != Gatherer.defaultInitializer()) // Optimization
+                state = initializer.get();
+            sink.begin(size);
+        }
+
+        @Override
+        public void accept(T t) {
+            /* Benchmarks have indicated that doing an unconditional write to
+             * `proceed` is more efficient than branching.
+             * We use `&=` here to prevent flips from `false` -> `true`.
+             *
+             * As of writing this, taking `greedy` or `stateless` into
+             * consideration at this point doesn't yield any performance gains.
+             */
+            proceed &= integrator.integrate(state, t, this);
+        }
+
+        @Override
+        public boolean cancellationRequested() {
+            return cancellationRequested(proceed);
+        }
+
+        private boolean cancellationRequested(boolean knownProceed) {
+            return !(knownProceed && (!sink.cancellationRequested() || (proceed = false)));
+        }
+
+        @Override
+        public void end() {
+            final var finisher = gatherer.finisher();
+            if (finisher != Gatherer.<A, R>defaultFinisher()) // Optimization
+                finisher.accept(state, this);
+            sink.end();
+            state = null; // GC assistance
```

Comment:
Not sure it really help the GC, given that it may go through a GC barrier

Reply:
When it helps, it helps, and when it doesn't it has negligible cost.